### PR TITLE
feat: team locks backend implementation

### PR DESCRIPTION
**Pull request by miguel-crespo-fdc**
Backend implementation of team locks. 
PR for frontend: https://github.com/freiheit-com/kuberpult/pull/1497

### CODE REVIEW

File: services/cd-service/pkg/repository/repository.go [hunk 1]
```diff
@@ -1517,7 +1519,25 @@ func (s *State) GetEnvironmentApplicationLocks(environment, application string)
...
-
+func (s *State) GetEnvironmentTeamLocks(environment, team string) (map[string]Lock, error) {
+	base := s.GetTeamLocksDir(environment, team)
+	if entries, err := s.Filesystem.ReadDir(base); err != nil {
+		return nil, err
+	} else {
+		result := make(map[string]Lock, len(entries))
+		for _, e := range entries {
+			if !e.IsDir() {
+				return nil, fmt.Errorf("error getting application locks: found file in the locks directory. run migration script to generate correct metadata")
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/repository.go:1530]:
```suggestion
				return nil, fmt.Errorf("error getting team locks: found file in the locks directory. run migration script to generate correct metadata")
```
Also, what is the migration script?

File: services/cd-service/pkg/repository/transformer.go [hunk 2]
```diff
@@ -1524,6 +1524,107 @@ func (c *DeleteEnvironmentApplicationLock) Transform(
...
+
+var ErrTeamNotFound error
+
+func (c *CreateEnvironmentTeamLock) Transform(
+	ctx context.Context,
+	state *State,
+	t TransformerContext,
+) (string, error) {
+	// Note: it's possible to lock an application BEFORE it's even deployed to the environment.
+	err := state.checkUserPermissions(ctx, c.Environment, "*", auth.PermissionCreateLock, c.Team, c.RBACConfig)
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer.go:1543]: we need to do this check for applications of the given team

File: services/cd-service/pkg/repository/transformer.go [hunk 3]
```diff
@@ -1524,6 +1524,107 @@ func (c *DeleteEnvironmentApplicationLock) Transform(
...
+					foundTeam = true
+					break
+				}
+			} else {
+				logger.FromContext(ctx).Sugar().Warnf("CreateEnvironmentTeamLock: Could not find team for application: %s.", currentApp)
+			}
+		}
+	}
+	if err != nil || !foundTeam { //Not found team or apps dir doesn't exist
+		return "", grpc.FailedPrecondition(ctx, fmt.Errorf("Team %s does not exist.", c.Team))
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer.go:1564]:
```suggestion
		return "", grpc.FailedPrecondition(ctx, fmt.Errorf("Team '%s' does not exist.", c.Team))
```

File: services/cd-service/pkg/repository/transformer.go [hunk 4]
```diff
@@ -1524,6 +1524,107 @@ func (c *DeleteEnvironmentApplicationLock) Transform(
...
+	}
+
+	envDir := fs.Join("environments", c.Environment)
+	if _, err := fs.Stat(envDir); err != nil {
+		return "", fmt.Errorf("error accessing dir %q: %w", envDir, err)
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer.go:1569]:
```suggestion
		return "", fmt.Errorf("error environment not found %q: %w", envDir, err)
```

File: services/cd-service/pkg/repository/transformer.go [hunk 5]
```diff
@@ -1524,6 +1524,107 @@ func (c *DeleteEnvironmentApplicationLock) Transform(
...
+	}
+
+	teamDir := fs.Join(envDir, "teams", c.Team)
+	if err := fs.MkdirAll(teamDir, 0777); err != nil {
+		return "", err
+	}
+	chroot, err := fs.Chroot(teamDir)
+	if err != nil {
+		return "", err
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer.go:1578]: same as above

File: services/cd-service/pkg/repository/transformer.go [hunk 6]
```diff
@@ -1524,6 +1524,107 @@ func (c *DeleteEnvironmentApplicationLock) Transform(
...
+	}
+	if err := createLock(ctx, chroot, c.LockId, c.Message); err != nil {
+		return "", err
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer.go:1581]: same as above

File: services/cd-service/pkg/repository/transformer.go [hunk 7]
```diff
@@ -1524,6 +1524,107 @@ func (c *DeleteEnvironmentApplicationLock) Transform(
...
+	}
+	// locks are invisible to argoCd, so no changes here
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer.go:1583]: I don't understand this comment

File: services/cd-service/pkg/repository/transformer.go (hunk 2 above)
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/transformer.go:1543]: ah, nevermind, we have the `team` parameter. Then this is fine.

File: services/cd-service/pkg/repository/transformer.go [hunk 8]
```diff
@@ -1649,7 +1750,20 @@ func (c *DeployApplicationVersion) Transform(
...
 			return "", err
 		}
-		if len(envLocks) > 0 || len(appLocks) > 0 {
+
+		appDir := applicationDirectory(fs, c.Application)
+
+		team, err := util.ReadFile(fs, fs.Join(appDir, "team"))
+
+		if errors.Is(err, os.ErrNotExist) {
+			teamLocks = map[string]Lock{} //This is a workaround. This transformer now needs a team file to be created on the application folder in order for the team locks to be loaded correctly. It turns out almost no other test that uses the CreateApplicationVersion transformer (that uses this one) sets a value for the team parameter, so the file wont exist and the tests will fail. IF we can't find the file, simply create an empty map and move on.
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer.go:1759]: this is not just a test issue, but will always happen when we create the first lock for a given team.
So I don't think this is a "workaround". If there's no directory, then there is no lock, that's working as expected.

File: services/cd-service/pkg/repository/transformer_test.go [hunk 9]
```diff
@@ -1708,6 +1708,75 @@ func TestApplicationDeploymentEvent(t *testing.T) {
...
+					},
+				},
+				&CreateEnvironment{
+					Environment: "staging",
+					Config: config.EnvironmentConfig{
+						Upstream: &config.EnvironmentConfigUpstream{
+							Latest: true,
+						},
+					},
+				},
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer_test.go:1729]: I think adding `staging` does not add anything critical for the test case.
Suggestion: remove staging environment.

File: services/cd-service/pkg/repository/transformer_test.go [hunk 10]
```diff
@@ -3974,6 +4086,138 @@ func TestRbacTransformerTest(t *testing.T) {
...
+					Environment: "production",
+					Team:        "sre-team",
+					Message:     "don't",
+					LockId:      "manual",
+					Authentication: Authentication{RBACConfig: auth.RBACConfig{DexEnabled: true, Policy: map[string]*auth.Permission{
+						"developer,CreateLock,production:production,*,allow": {Role: "Developer"},
+					}}},
+				},
+			},
+		},
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer_test.go:4147]: I assume `*` here means "any team".
In that case: Please add another test that checks that it works for "sre-team" as well.

File: services/cd-service/pkg/service/batch.go [hunk 11]
```diff
@@ -147,6 +165,30 @@ func (d *BatchServer) processAction(
 			LockId:         act.LockId,
 			Authentication: repository.Authentication{RBACConfig: d.RBACConfig},
 		}, nil, nil
+	case *api.BatchAction_CreateEnvironmentTeamLock:
+		act := action.CreateEnvironmentTeamLock
+
+		if err := ValidateEnvironmentTeamLock("create", act.Environment, act.Team, act.LockId); err != nil {
+			return nil, nil, err
+		}
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/service/batch.go:173]: this whole function is supposed to be deprecated. 
Please put the validations into the actual transformer.

File: services/cd-service/pkg/service/batch.go [hunk 12]
```diff
@@ -147,6 +165,30 @@ func (d *BatchServer) processAction(
...
+		return &repository.CreateEnvironmentTeamLock{
+			Environment:    act.Environment,
+			Team:           act.Team,
+			LockId:         act.LockId,
+			Message:        act.Message,
+			Authentication: repository.Authentication{RBACConfig: d.RBACConfig},
+		}, nil, nil
+	case *api.BatchAction_DeleteEnvironmentTeamLock:
+		act := action.DeleteEnvironmentTeamLock
+		if err := ValidateEnvironmentTeamLock("delete", act.Environment, act.Team, act.LockId); err != nil {
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/service/batch.go:183]: same about validations

File: services/cd-service/pkg/service/batch_test.go [hunk 13]
```diff
@@ -277,11 +311,47 @@ func TestBatchServiceWorks(t *testing.T) {
...
+				_, exists = teamLocks["91011"]
+				if exists {
+					t.Error("lock was not deleted")
+				}
+			}
 
 		})
 	}
 }
+func listFilesHelper(fs billy.Filesystem, path string) []string {
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/service/batch_test.go:336]: if this is unused, please remove the function.

File: services/cd-service/pkg/service/overview.go [hunk 14]
```diff
@@ -147,6 +150,27 @@ func (o *OverviewServiceServer) getOverview(
 							DeployTime:   "",
 						},
 					}
+					//we don't want to insert the locks for a team twice, so init a set of checked teams and filter the ones we already processed.
+					teamName, err := s.GetTeamName(appName)
+					_, exists := teamSet[teamName]
+					if err == nil && !exists {
+						teamSet[teamName] = true
+						if teamLocks, teamErr := s.GetEnvironmentTeamLocks(envName, teamName); teamErr == nil {
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/service/overview.go:158]: What if there is a `teamErr`?
It's reasonable to still return something to the UI here, but we should at least log a warning.

File: services/cd-service/pkg/repository/transformer_test.go (hunk 10 above)
Comment (miguel-crespo-fdc) [reply services/cd-service/pkg/repository/transformer_test.go:4147]: It is for the application, not the team

Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/transformer_test.go:4137]: then ignore my comment, I misunderstood

**Pull request merged**